pytorch_lightning/utilities/apply_func.py

```diff
@@ -222,6 +222,47 @@ def apply_to_collections(
         ]
         return elem_type(*out) if is_namedtuple else elem_type(out)
 
+    if _is_dataclass_instance(data1):
+        assert isinstance(data2, type(data1)), "Dataclasses are of different types"
```

Comment:
Should we assert the dataclasses have the same fields instead ?

Reply:
sure, I can do that instead. 
but I wonder, is it an expected situation where the user has defined two different types with the same fields and attempts to merge them together?
and, what should the type of reduced result be?